src/vec.rs

```diff
@@ -53,16 +56,88 @@ pub trait FallibleVec<T> {
         T: TryClone;
 }
 
+#[cfg(not(feature = "unstable"))]
+fn vec_try_reserve<T>(v: &mut Vec<T>, additional: usize) -> Result<(), TryReserveError> {
+    let available = v.capacity().checked_sub(v.len()).expect("capacity >= len");
+    if additional > available {
+        let increase = additional
+            .checked_sub(available)
+            .expect("additional > available");
+        let new_cap = v
+            .capacity()
+            .checked_add(increase)
+            .ok_or(TryReserveError::CapacityOverflow)?;
+        vec_try_extend(v, new_cap)?;
+        debug_assert!(v.capacity() == new_cap);
+    }
+
+    Ok(())
+}
+
+#[cfg(not(feature = "unstable"))]
+fn vec_try_extend<T>(v: &mut Vec<T>, new_cap: usize) -> Result<(), TryReserveError> {
+    let old_len = v.len();
+    let old_cap: usize = v.capacity();
+
+    if old_cap >= new_cap {
+        return Ok(());
+    }
+
+    let elem_size = core::mem::size_of::<T>();
+    let new_alloc_size = new_cap
+        .checked_mul(elem_size)
+        .ok_or(TryReserveError::CapacityOverflow)?;
+
+    // required for alloc safety
+    // See https://doc.rust-lang.org/stable/std/alloc/trait.GlobalAlloc.html#safety-1
+    // Should be unreachable given prior `old_cap >= new_cap` check.
+    assert!(new_alloc_size > 0);
```

Comment:
This one, I think we should leave alone since I believe it's impossible to reach this point unless `new_alloc_size` is greater than zero. `Vec::capacity` for ZSTs returns `usize::MAX`, so we'll always return early. I'll add some additional tests to exercise this bit, but I think `Vec` and `TryVec` should work just fine with ZSTs.